Account Mutations for GraphQL

shomykohai opened the pull request:
# What

<!-- Please summarize your changes.  This includes:

- What bugs did you fix?
- What features did you add?
- Did you add any dependencies? -->

Added mutations related to Account & Profile of a User:
- Account mutations (name, slug, sfw filter)
- Profile mutations (birthday, gender, about..)
- Profile Links
- Favorites (Characters, Anime, Manga..)
- Blocks (Block or unblock someone)

# Why

<!-- Explain why you implemented this how you did.  This includes:

- What decisions did you make while building this?
- What are some alternatives you considered?
- What are some downsides of the approach you chose?
- Did you discuss those decisions with anyone else?  What were their thoughts? -->

To make this mutations avaiable on graphql for V4

# Checklist

<!-- ALL PULL REQUESTS -->
- [x] All files pass Rubocop
- [x] Any complex logic is commented
- [x] Any new systems have thorough documentation
- [x] Any user-facing changes are behind a feature flag (or: explain why they can't be)
<!-- FINISHED (NON-DRAFT) PULL REQUESTS -->
- [x] All the tests pass
- [x] Tests have been added to cover the new code

shomykohai committed:
Account Mutation

shomykohai committed:
More mutations

Made mutations only for account and profile type

shomykohai committed:
Profile links mutations

shomykohai committed:
Make Staff able to edit user profile

shomykohai committed:
Profile mutations

Made profile mutations not part of account mutations (excluding the Update of Account)

shomykohai committed:
Merge branch 'hummingbird-me:the-future' into account-mutations

shomykohai committed:
Add Favorite mutations

* Adds favorite entry mutations:
- Anime
- Manga
- Character

shomykohai committed:
Add Block mutations

Add mutations to GraphQL to block and unblock a user

shomykohai committed:
Fix for checks

NuckChorris commented:
Do you think you could refactor these on top of the `FancyMutation` stuff like I did for [the MediaReaction mutations](https://github.com/hummingbird-me/kitsu-server/blob/the-future/app/graphql/mutations/media_reaction/create.rb)? Should make the code a lot nicer, it handles all the common structural stuff, pulls the input and output types into the mutation file, and provides a standard for how to deal with errors!

It should be pretty easy, you can splat the keyword params from the input object into the user like this:

```ruby
def resolve(**input)
  current_user.update!(**input)
end
```

or manually assign params through, like this:

```ruby
def resolve(blocked_id:)
  Block.create!(
    user: current_user,
    blocked_id:
  )
end
```

Also, it's now super easy to write tests for graphql stuff, using `with graphql helpers` as you can see [here](https://github.com/hummingbird-me/kitsu-server/blob/the-future/spec/graphql/mutations/media_reaction/create_spec.rb), so if you could add tests that would be ideal, so we know everything works as it should 🎉

NuckChorris commented:
Oh yeah also if you wouldn't mind, the profile mutations should probably be in a separate Profile namespace. The GraphQL API tries to maintain a solid separation between "Account" (private) and "Profile" (public) stuff 😅

shomykohai committed:
Resolve merge conflicts

shomykohai committed:
Merge branch 'account-mutations' of https://github.com/ShomyKohai/kitsu-server into account-mutations

shomykohai committed:
Revert resolve merge conflicts

shomykohai committed:
Merge branch 'hummingbird-me:the-future' into account-mutations

shomykohai committed:
Merge branch 'hummingbird-me:the-future' into account-mutations

shomykohai committed:
Refactor Account mutations

Refactor Account mutations on top of FancyMutation

shomykohai committed:
Fix NotFound errors

shomykohai committed:
Add Validation Error to ProfileLinks

shomykohai committed:
Build NotAuthenticated in ProfileLink::Create

shomykohai committed:
Fix NotFound Error in ProfileLink::Update

shomykohai committed:
FIx more NotFound Errors

shomykohai committed:
Add Tests

Add test to
* Profile (update)
* Account (update)
* Block (create & delete)
* Favorite (create & delete)
* ProfileLink (create, update & delete)

NuckChorris commented on app/graphql/mutations/account/update.rb:
```diff
@@ -0,0 +1,58 @@
+  description 'Update account'
+
+  input do
+    argument :name,
+      String,
+      required: false,
+      description: 'The display name of the user'
+    argument :slug,
+      String,
+      required: false,
+      description: 'The slug (@username) of the user'
```
These should probably be on the `Profile::Update` since they're not on the `Account` type!

shomykohai replied:
I was thinking the same, but then I saw how in the client these fields are in the account settings

![image](https://github.com/hummingbird-me/kitsu-server/assets/61943525/385fdf78-bb35-4d78-917c-c3c3679cfdf8)

Are these option in the new react client gonna be moved to the profile page or it does not matter (regarding on how the client ask the changes to the api)?

NuckChorris commented on app/graphql/mutations/block/create.rb:
```diff
@@ -0,0 +1,32 @@
+      user_id: current_user.id,
+      blocked_id:
+    )
+    authorize!(@block, :create?)
```
You need to add `NotAuthorized` as a potential error, and probably want to move this up into the `ready?` method

NuckChorris commented on app/graphql/mutations/favorite/create.rb:
```diff
@@ -0,0 +1,46 @@
+      item_id:,
+      user_id: current_user.id
+    )
+    authorize!(@favorite, :create?)
```
Again, probably want to put this in `ready?` and add `NotAuthorized` as a possible error!

NuckChorris commented on app/graphql/mutations/profile/update.rb:
```diff
@@ -0,0 +1,47 @@
+  description 'Update profile'
+
+  input do
+    argument :id, ID,
+      required: true,
+      description: 'Your ID or the one of another user.'
```
I think let's make this optional, it's only useful to admins, and the default can just be `current_user`

NuckChorris commented on app/graphql/mutations/profile_link/create.rb:
```diff
@@ -0,0 +1,46 @@
+      profile_link_site_id: profile_link,
+      user_id: current_user.id
+    )
+    raise GraphQL::ExecutionError, 'You already have this profile link.' if profile_link
```
A better way to handle this is probably to add a `validates :profile_link_site, uniqueness: { scope: :user }` on the ProfileLink model, and rely on `Types::Errors::Validation` to handle this

NuckChorris commented on app/graphql/mutations/profile_link/delete.rb:
```diff
@@ -0,0 +1,30 @@
+  description 'Delete a profile link'
+
+  input do
+    argument :profile_link_id, ID,
+      required: true,
+      description: 'The profile link to delete'
```
Using id here vs ProfileLinkSite in update makes me think we should just choose one or the other to "identify" the link. I'm kinda favoring just ProfileLinkSite, so that clients don't have to find the ID before deletion

NuckChorris commented on app/graphql/types/block.rb:
```diff
@@ -0,0 +1,20 @@
+    null: false,
+    description: 'User who got blocked.'
+
+  def blocked_user
+    Loaders::RecordLoader.for(User).load(object.blocked_id)
+  end
```
You probably also want this for `user` (not that it's likely to ever be navigated the other direction, really just in case)

NuckChorris commented on app/graphql/types/query_type.rb:
```diff
@@ -9,6 +9,14 @@ def current_account
   end
 
+  field :current_profile, Types::Profile, null: true do
+    description 'Kitsu account details. You must supply an Authorization token in header.'
```
Could you change the wording here?

```suggestion
    description 'Your Kitsu profile. You must supply an Authorization token in header.'
```

NuckChorris commented on app/graphql/types/enum/favorite_item.rb:
```diff
@@ -0,0 +1,6 @@
+class Types::Enum::FavoriteItem < Types::Enum::Base
```
Hmmm, could you rename this to `FavoriteType`? I think it would better express what this is, and allow us to use it elsewhere (ie, in queries like `user { favorites(type: ANIME) }`

shomykohai committed:
Add NotAuthorized to Block

shomykohai committed:
Rename FavoriteItem to FavoriteType

shomykohai committed:
Make id for profile update mutation optional

shomykohai committed:
Use ProfileLinkSite instead of Id in ProfileLink::Delete

shomykohai committed:
Add validations to ProfileLinks

Add validation to ProfileLinks so we don't need to check the uniqueness in the GraphQL mutation

shomykohai committed:
Load user with RecordLoader in Types::Block

shomykohai committed:
Fix`current_profile` description

shomykohai committed:
Move `name` and `slug` to Profile::Update

NuckChorris commented on app/graphql/mutations/profile/update.rb:
```diff
@@ -0,0 +1,59 @@
+
+  def resolve(**input)
+    @profile.update!(**input)
+    @profile.tap(&:save!)
```
This saves twice, you probably want to use `assign_attributes` and then `save!` or just only use `update!`

Here's a cheat sheet on these methods:

https://scottbartell.com/2020/01/30/set-attributes-in-active-record-rails-6/

NuckChorris commented on app/graphql/types/query_type.rb:
```diff
@@ -442,4 +450,21 @@ def reports_by_status(statuses: nil, sort: [{ on: :created_at, direction: :asc }
       sort: sort
     }, statuses)
   end
+
+  field :find_block_by_id, Types::Block, null: true do
+    description 'Find a single Block by ID'
+    argument :id, ID, required: true
+  end
+
+  def find_block_by_id(id:)
+    Loaders::RecordLoader.for(::Block, token: context[:token]).load(id)
+  end
+
```
Is this actually useful for any purpose? We should only add fields that we know will be useful, because it's a lot easier to add something than remove it 😉

shomykohai replied:
When I first added this I thought about some use (that unfortunately i don't remember atm).
But i agree that now it does not have purpose or use, I'll remove it!

NuckChorris reviewed (commented):
Some minor comments, but otherwise it looks good and I think we should be ready to merge after this!

I'm working on changing the `authenticate!` and `authorize!` methods to use a raised exception to end execution, but this will all continue to work so we can merge this first and I can change that later :+1:

shomykohai committed:
Stop record from get saved twice

shomykohai committed:
Remove findBlockById query

shomykohai commented:
Just made the changes!
I also removed the double save from profile links

NuckChorris reviewed (approved):
Looks great, let's gooooo!
